lib/utils/pdf_to_text.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+class DiscourseAi::Utils::PdfToText
+  class Reader
+    def initialize(upload:, user: nil, llm_model: nil)
+      @extractor =
+        DiscourseAi::Utils::PdfToText.new(upload: upload, user: user, llm_model: llm_model)
+      @enumerator = create_enumerator
+      @buffer = +""
+    end
+
+    def read(length)
+      return @buffer.slice!(0, length) if !@buffer.empty?
+
+      begin
+        @buffer << @enumerator.next
+      rescue StopIteration
+        return nil
+      end
+
+      @buffer.slice!(0, length)
+    end
+
+    private
+
+    def create_enumerator
+      Enumerator.new { |yielder| @extractor.extract_text { |chunk| yielder.yield(chunk || "") } }
+    end
+  end
+
+  attr_reader :upload
+
+  def self.as_fake_file(upload:, user: nil, llm_model: nil)
+    Reader.new(upload: upload, user: user, llm_model: llm_model)
+  end
+
+  def initialize(upload:, user: nil, llm_model: nil)
+    @upload = upload
+    @user = user
+    @llm_model = llm_model
+  end
+
+  def extract_text
+    pdf_path =
+      if upload.local?
+        Discourse.store.path_for(upload)
+      else
+        Discourse.store.download_safe(upload, max_file_size_kb: MAX_PDF_SIZE)&.path
```

Comment:
The `MAX_PDF_SIZE ` constant is not defined in this class so this will blow up when S3 uploads are enabled. Should we add a test to cover this change here?

Reply:
oh yikes, will fix